src/modules/flow/iio/nodes.c

```diff
@@ -25,43 +25,117 @@
+    struct sol_flow_node_type base;
+    uint16_t out_port;
+};
+
 static void
 reader_cb(void *data, struct sol_iio_device *device)
 {
     static const char *errmsg = "Could not read channel buffer values";
     struct sol_flow_node *node = data;
-    struct gyroscope_data *mdata = sol_flow_node_get_private_data(node);
-    struct sol_direction_vector out = {
-        .min = mdata->out_range.min,
-        .max = mdata->out_range.max
-    };
+    struct iiodevice_data *mdata = sol_flow_node_get_private_data(node);
+    struct iiodevice_out out;
+    struct iio_node_type *type;
+    double tmp;
     bool b;
 
-    b = sol_iio_read_channel_value(mdata->channel_x, &out.x);
-    if (!b) goto error;
-
-    b = sol_iio_read_channel_value(mdata->channel_y, &out.y);
-    if (!b) goto error;
-
-    b = sol_iio_read_channel_value(mdata->channel_z, &out.z);
-    if (!b) goto error;
+    type = (struct iio_node_type *)sol_flow_node_get_type(node);
+
+    switch (mdata->type) {
+        case IIO_GYROSCOPE:
+        case IIO_MAGNETOMETER:
+        case IIO_ACCELEROMETER:
+            out.direction.min = mdata->out_range.min;
+            out.direction.max = mdata->out_range.max;
+            b = sol_iio_read_channel_value(mdata->channels[0], &out.direction.x);
+            if (!b) goto error;
+
+            b = sol_iio_read_channel_value(mdata->channels[1], &out.direction.y);
+            if (!b) goto error;
+
+            b = sol_iio_read_channel_value(mdata->channels[2], &out.direction.z);
+            if (!b) goto error;
+            break;
+        case IIO_COLOR:
+            out.rgb.red_max= mdata->out_range.max;
+            out.rgb.green_max = mdata->out_range.max;
+            out.rgb.blue_max = mdata->out_range.max;
+            b = sol_iio_read_channel_value(mdata->channels[0], &tmp);
+            if (!b || tmp < 0 || tmp > UINT32_MAX) goto error;
+            out.rgb.red = tmp;
+
+            b = sol_iio_read_channel_value(mdata->channels[1], &tmp);
+            if (!b || tmp < 0 || tmp > UINT32_MAX) goto error;
+            out.rgb.green = tmp;
+
+            b = sol_iio_read_channel_value(mdata->channels[2], &tmp);
+            if (!b || tmp < 0 || tmp > UINT32_MAX) goto error;
+            out.rgb.blue = tmp;    
+            break;
+         case IIO_THERMOMETER:
+         case IIO_PRESSURE_SENSOR:
+         case IIO_HUMIDITY_SENSOR:
+         case IIO_ADC:
+         case IIO_LIGHT_SENSOR:
+         case IIO_PROXIMITY:
+            out.drange.min = mdata->out_range.min;
+            out.drange.max = mdata->out_range.max;
+            out.drange.step = mdata->out_range.step;
+            b = sol_iio_read_channel_value(mdata->channels[0], &out.drange.val);
+            if (!b) goto error;
+            break;
+        default:
+            break;
+    }
 
-    sol_flow_send_direction_vector_packet(node,
-        SOL_FLOW_NODE_TYPE_IIO_GYROSCOPE__OUT__OUT, &out);
+    sol_flow_send_direction_vector_packet(node, type->out_port, &out.direction);
```

Comment:
Why are you always sending a direction-vector packet, even when port is of a different packet type?